Modules/resource.c

```diff
@@ -340,144 +340,129 @@ resource_methods[] = {
 /* Module initialization */
 
 
-static struct PyModuleDef resourcemodule = {
-    PyModuleDef_HEAD_INIT,
-    "resource",
-    NULL,
-    -1,
-    resource_methods,
-    NULL,
-    NULL,
-    NULL,
-    NULL
-};
-
-PyMODINIT_FUNC
-PyInit_resource(void)
+static int resource_exec(PyObject *module)
 {
-    PyObject *m, *v;
-
-    /* Create the module and add the functions */
-    m = PyModule_Create(&resourcemodule);
-    if (m == NULL)
-        return NULL;
+    PyObject *v;
 
     /* Add some symbolic constants to the module */
     Py_INCREF(PyExc_OSError);
-    PyModule_AddObject(m, "error", PyExc_OSError);
+    if (PyModule_AddObject(module, "error", PyExc_OSError) < 0) {
+        Py_DECREF(PyExc_OSError);
+        return -1;
+    }
     if (!initialized) {
         if (PyStructSequence_InitType2(&StructRUsageType,
                                        &struct_rusage_desc) < 0)
-            return NULL;
+            return -1;
     }
 
-    Py_INCREF(&StructRUsageType);
-    PyModule_AddObject(m, "struct_rusage",
-                       (PyObject*) &StructRUsageType);
+    if(PyModule_AddType(module, &StructRUsageType) < 0) {
+        return -1;
+    }
 
     /* insert constants */
 #ifdef RLIMIT_CPU
-    PyModule_AddIntMacro(m, RLIMIT_CPU);
+    PyModule_AddIntMacro(module, RLIMIT_CPU);
```

Comment:
While you modify the code, would you mind to add error handling? I suggest to add a macro which handles error. Something like:

```
#define ADD_INT(modle, value) \
    do { \
        if (PyModule_AddIntMacro(module, value) < 0) { \
            return -1; \
        } \
    } while (0)

    /* insert constants */
#ifdef RLIMIT_CPU
    ADD_INT(module, RLIMIT_CPU);
#endif
    (...)
```

Reply:
Good idea ;) I plan use a seperate bpo to add this macro.

Reply:
I don't think that fixing such bug requires a separated bpo. You can add the macro in this PR. I would prefer to not push 3 commits per module to port them to multiphase initialization and fix reference leaks in the error path. I would prefer 1 commit per module.

Reply:
Extension module like `_locale` called `PyModule_AddIntMacro()` too(which have ported to multiphase initialization already)， so my intuitive response is:
1. Adding a global macro like `ADD_INT`;
2. some ported extension module use `ADD_INT`;
3. this PR use `ADD_INT` finally;

but it's make no difference, right?

Reply:
Please don't make such macro public. It must remains private, and only be declared in each .c file. I hate macros using "goto". But if it's only used in one function and declared inside the function, it's fine.

Reply:
Adding PyModule_AddType() was different: it returns -1 on failure, it's a function, not a macro.

Reply:
Got it, thank you for your detailed explanation.